token/services/ttx/support.go

```diff
@@ -54,3 +57,128 @@ func StoreEnvelope(context view.Context, tx *Transaction) error {
+		if err != nil {
+			logger.Errorf("failed to run view: %s", err)
+		}
+	}()
+}
+
+// LocalBidirectionalChannel is a bidirectional channel that is used to simulate
+// a session between two views (let's call them L and R) running in the same process.
+type LocalBidirectionalChannel struct {
+	left  view.Session
+	right view.Session
+}
+
+// NewLocalBidirectionalChannel creates a new bidirectional channel
+func NewLocalBidirectionalChannel(caller string, contextID string, endpoint string, pkid []byte) (*LocalBidirectionalChannel, error) {
+	ID, err := comm.GetRandomNonce()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to generate session ID")
+	}
+	lr := make(chan *view.Message, 10)
+	rl := make(chan *view.Message, 10)
+
+	info := view.SessionInfo{
+		ID:           base64.StdEncoding.EncodeToString(ID),
+		Caller:       nil,
+		CallerViewID: "",
+		Endpoint:     endpoint,
+		EndpointPKID: pkid,
+		Closed:       false,
+	}
+	return &LocalBidirectionalChannel{
+		left: &localSession{
+			name:         "left",
+			contextID:    contextID,
+			caller:       caller,
+			info:         info,
+			readChannel:  rl,
+			writeChannel: lr,
+		},
+		right: &localSession{
+			name:         "right",
+			contextID:    contextID,
+			caller:       caller,
+			info:         info,
+			readChannel:  lr,
+			writeChannel: rl,
+		},
+	}, nil
+}
+
+// LeftSession returns the session from the L to R
+func (c *LocalBidirectionalChannel) LeftSession() view.Session {
+	return c.left
+}
+
+// RightSession returns the session from the R to L
+func (c *LocalBidirectionalChannel) RightSession() view.Session {
+	return c.right
+}
+
+// localSession is a local session that is used to simulate a session between two views.
+// It has a read channel and a write channel.
+type localSession struct {
+	name         string
+	contextID    string
+	caller       string
+	info         view.SessionInfo
+	readChannel  chan *view.Message
+	writeChannel chan *view.Message
+}
+
+func (s *localSession) Info() view.SessionInfo {
+	return s.info
+}
+
+func (s *localSession) Send(payload []byte) error {
+	logger.Debugf("[%s] Sending message to self session of length %d", s.name, len(payload))
+	s.writeChannel <- &view.Message{
+		SessionID:    s.info.ID,
+		ContextID:    s.contextID,
+		Caller:       s.caller,
+		FromEndpoint: s.info.Endpoint,
+		FromPKID:     s.info.EndpointPKID,
+		Status:       view.OK,
+		Payload:      payload,
+	}
+	return nil
+}
+
+func (s *localSession) SendError(payload []byte) error {
+	logger.Debugf("[%s] Sending error message to self session of length %d", s.name, len(payload))
+	s.writeChannel <- &view.Message{
+		SessionID:    s.info.ID,
+		ContextID:    s.contextID,
+		Caller:       s.caller,
+		FromEndpoint: s.info.Endpoint,
+		FromPKID:     s.info.EndpointPKID,
+		Status:       view.ERROR,
+		Payload:      payload,
+	}
+	return nil
+}
+
+func (s *localSession) Receive() <-chan *view.Message {
+	return s.readChannel
+}
+
+func (s *localSession) Close() {
```

Comment:
is this somehow used? For instances as check if Send Or Receive is called?

Reply:
let me update the code to check that flag.